### DISCUSSION TITLE
Out of memory

### DISCUSSION THREAD

**Discussion by AckslD** [kkharji/sqlite.lua#109] [Q&A] [answered]
Hi!

I'm trying out a minimal example to see how sqlite could be used in [neoclip](https://github.com/AckslD/nvim-neoclip.lua) but getting the error `"out of memory"` when opening the database. Any idea why?

```lua
local has_sqlite, sqlite = pcall(require, "sqlite")
if not has_sqlite then
    print "Couldn't find sqlite.lua. Cannot use persistent history"
end

local db = sqlite.new('test.sql')
print(vim.inspect(db:status()))
```
which prints
```
{
  code = 7,
  msg = "out of memory"
}
```

**Comment by Conni2461** [comment]
Status called instantly `sqlite3_status`  and puts the result in a table.

```
#define SQLITE_NOMEM        7   /* A malloc() failed */
```

We are probably getting the result here because `sqlite.new` doesnt actually create a new database yet just a logical lua object. He pushes that back till you actually do a insert/update/select/delete. (He had a little discussion where i said thats kinda weird behavior). 

So i guess sqlite3_status returns 7 if it wasnt actually "opened", "connected" to a database (memory or file).

`sqlite.open` actually opens it. If i recall correctly.

Edit: Correctly myself here a little bit (been thinking about it).

Because sqlite.new defers the actual `sqlite3_open` call (which returns the connection) status will be called here with null for the connection. So the return from `sqlite3_status` makes sense to me.

**Comment by kkharji** [comment, marked as answer]
Thanks @Conni2461, I was like how why this happened, 2000 lines of tests for nothing 🤣

Yes, `.new` or `.__call` just creates a logical object that doesn't even check if db file exists or anything. 

`.status` is a extremely low level api, meant for debugging and internal stuff. 

`.open` does what you would expect and `.status` should work and return code 0.

So you can either run `sqlite:open(path,...)` or with `.new` run `db:open()` first before doing any sqlite stuff (other than creating self-contained tables).

neoclip sounds really interesting, I'm thrilled to use it with sqlite. Tag me if you need help and good luck

**Reply by AckslD** [reply]
Thanks @Conni2461 and @tami5! That make sense and indeed after `open` the status is as expected :)

I have now a first implementation to have persistent history between sessions for neoclip. When it's finished I'll tag you in the PR @tami5. If you have time I'd be very happy to hear any suggestions if I'm using sqlite in a non-optimal way :)

**Reply by AckslD** [reply]
@tami5 I had another issue and tagged you in the PR. Feel free to let me know it you know what I'm doing wrong :) Or @Conni2461 if you're also interested:
https://github.com/AckslD/nvim-neoclip.lua/pull/20/files#r702087824
